lib/syscall_shim/include/uk/syscall.h

```diff
@@ -351,6 +367,250 @@ typedef long uk_syscall_arg_t;
+		"/* Align to `struct uk_syscall_regs` alignment */\n\t"	\
+		"bic	x0, x0, #"					\
+			STRINGIFY(UK_SYSCALL_REGS_END_ALIGN - 1)"\n\t"	\
+		"/* Make room for `struct uk_syscall_regs` */\n\t"	\
+		"sub	x0, x0, #"STRINGIFY(UK_SYSCALL_REGS_SIZE)"\n\t"	\
+		"/* Swap x0 and (old) sp */\n\t"			\
+		"add	sp, sp, x0\n\t"					\
+		"sub	x0, sp, x0\n\t"					\
+		"sub	sp, sp, x0\n\t"					\
+		"/* Now store old sp w.r.t. `struct __regs` */\n\t"	\
+		"str	x0, [sp, #"STRINGIFY(__SP_OFFSET)"]\n\t"	\
+		"/* Restore x0 from scratch register TPIDRRO_EL0 */\n\t"\
+		"mrs	x0, tpidrro_el0\n\t"				\
+		"/* Now just store the rest of `struct __regs` */\n\t"	\
+		"stp	x0, x1, [sp, #16 * 0]\n\t"			\
+		"stp	x2, x3, [sp, #16 * 1]\n\t"			\
+		"stp	x4, x5, [sp, #16 * 2]\n\t"			\
+		"stp	x6, x7, [sp, #16 * 3]\n\t"			\
+		"stp	x8, x9, [sp, #16 * 4]\n\t"			\
+		"stp	x10, x11, [sp, #16 * 5]\n\t"			\
+		"stp	x12, x13, [sp, #16 * 6]\n\t"			\
+		"stp	x14, x15, [sp, #16 * 7]\n\t"			\
+		"stp	x16, x17, [sp, #16 * 8]\n\t"			\
+		"stp	x18, x19, [sp, #16 * 9]\n\t"			\
+		"stp	x20, x21, [sp, #16 * 10]\n\t"			\
+		"stp	x22, x23, [sp, #16 * 11]\n\t"			\
+		"stp	x24, x25, [sp, #16 * 12]\n\t"			\
+		"stp	x26, x27, [sp, #16 * 13]\n\t"			\
+		"stp	x28, x29, [sp, #16 * 14]\n\t"			\
+		"mrs	x21, elr_el1\n\t"				\
+		"stp	x30, x21, [sp, #16 * 15]\n\t"			\
+		"mrs	x22, spsr_el1\n\t"				\
+		"mrs	x23, esr_el1\n\t"				\
+		"stp	x22, x23, [sp, #16 * 16]\n\t"			\
+		"/* ECTX at slot w.r.t. `struct uk_syscall_regs` */\n\t"\
+		"mov	x0, sp\n\t"					\
+		"add	x0, x0, #("STRINGIFY(__REGS_SIZEOF +		\
+				     UKARCH_ULCTX_SIZE)")\n\t"		\
+		"bl	ukarch_ectx_store\n\t"				\
+		"/* ULCTX at slot w.r.t. `struct uk_syscall_regs` */\n\t"\
+		"mov	x0, sp\n\t"					\
+		"add	x0, x0, #"STRINGIFY(__REGS_SIZEOF)"\n\t"	\
+		"bl	ukarch_ulctx_switchoff\n\t"			\
+		"mov	x0, sp\n\t"					\
+		"msr	daifclr, #2\n\t"				\
+		"bl	"STRINGIFY(fname)"\n\t"				\
+		"/* Only restore callee preserved regs (ABI) */\n\t"	\
+		"ldr	x30, [sp, #16 * 15]\n\t"			\
+		"ldp	x28, x29, [sp, #16 * 14]\t\n"			\
+		"ldp	x26, x27, [sp, #16 * 13]\t\n"			\
+		"ldp	x24, x25, [sp, #16 * 12]\t\n"			\
+		"ldp	x22, x23, [sp, #16 * 11]\t\n"			\
+		"ldp	x20, x21, [sp, #16 * 10]\t\n"			\
+		"ldp	x18, x19, [sp, #16 * 9]\t\n"			\
+		"/* Restore rsp from where it was stored */\n\t"	\
+		"ldr	x9, [sp, #"STRINGIFY(__SP_OFFSET)"]\n\t"	\
+		"mov	sp, x9\n\t"					\
+		"ret\n\t"						\
+		::							\
+		);							\
+									\
+		return 0;						\
+	}
+#else /* !__X86_64__ && !__ARM_64__ */
+#error "Undefined architecture"
+#endif
+
+#define __UK_LLSYSCALL_R_U_DEFINE(x, xa, rtype, name, ename, rname, ...)\
+	long rname(long _usr);						\
+	long __used ename(long _usr)						\
+	{								\
+		long ret;						\
+									\
+		ret = rname(_usr);					\
+		if (ret < 0 && PTRISERR(ret)) {				\
+			errno = -(int) PTR2ERR(ret);			\
+			return -1;					\
+		}							\
+		return ret;						\
+	}								\
+	static inline rtype __##rname(struct uk_syscall_regs *usr,	\
```

Comment:
This one is problematic in syscalls with no args like `vfork`, as it evaluates to:
```
__##rname(struct uk_syscall_regs *usr, )
```
To reproduce:
```
UK_LLSYSCALL_R_U_DEFINE(pid_t, vfork)
```